### ISSUE TITLE
make error on OSX

### ISSUE THREAD

**Issue by kathur2rain** [systemed/tilemaker#137]
since there's no make and install instruction for osx,  I guess I could simply type make in directory. after some process I get this error:

ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [tilemaker] Error 1

I'm using Mojave  10.14.4
thanks in advance

**Comment by systemed** [comment]
Works ok for me (also 10.14.4) - have you installed the dependencies?

**Comment by mdt-kathur** [comment]
yes I got /protobuf 3.7.1/ boost 1.70.0/ Lua@5.1 5.1.5_8/shapelib 1.5.0/ sqlite 3.28.0 have I miss something? or used the wrong version?

**Comment by systemed** [comment]
Don't know, then, I'm afraid.

**Comment by mdt-kathur** [comment]
@systemed  thanks for the reply~ here's some more information on the make error. 
g++ -O3 -Wall -Wno-unknown-pragmas -Wno-sign-compare -std=c++11 -pthread  -o tilemaker include/osmformat.pb.o include/vector_tile.pb.o clipper/clipper.o src/mbtiles.o src/pbf_blocks.o src/coordinates.o src/osm_store.o src/helpers.o src/output_object.o src/read_shp.o src/read_pbf.o src/osm_lua_processing.o src/write_geometry.o src/shared_data.o src/tile_worker.o src/tile_data.o src/osm_mem_tiles.o src/shp_mem_tiles.o src/tilemaker.o -I/usr/local/include -isystem ./include -I./src -I/usr/local/include/lua5.1 -I/usr/include/lua5.1 -L/usr/local/lib -lz -llua5.1 -lboost_program_options -lsqlite3 -lboost_filesystem -lboost_system -lprotobuf -lshp
Undefined symbols for architecture x86_64:
  "_luaL_loadfilex", referenced from:
      kaguya::State::dofile(char const*, kaguya::LuaTable const&) in osm_lua_processing.o
  "_luaL_traceback", referenced from:
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<void (*)(kaguya::ObjectWrapperBase*)> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (OsmLuaProcessing::*)() const> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<bool (OsmLuaProcessing::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (OsmLuaProcessing::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > (OsmLuaProcessing::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<bool (OsmLuaProcessing::*)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)> >, void>::invoke(lua_State*) in osm_lua_processing.o
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<bool (OsmLuaProcessing::*)() const> >, void>::invoke(lua_State*) in osm_lua_processing.o
      ...
  "_lua_absindex", referenced from:
      kaguya::util::stackValueDump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, lua_State*, int, int) in read_shp.o
      kaguya::FunctionResults kaguya::detail::LuaVariantImpl<kaguya::TableKeyReferenceProxy<char const*> >::operator()<OsmLuaProcessing*>(OsmLuaProcessing*&&) in osm_lua_processing.o
      kaguya::UserdataMetatable<OsmLuaProcessing, void>::createMatatable(lua_State*) const in osm_lua_processing.o
      kaguya::LuaStackRef kaguya::detail::LuaTableOrUserDataImpl<kaguya::LuaStackRef>::getRawField<char [7]>(char const (&) [7]) const in osm_lua_processing.o
      kaguya::class_userdata::newmetatable(lua_State*, std::type_info const&, char const*) in osm_lua_processing.o
      kaguya::LuaTable kaguya::detail::FunctionResultProxy::ReturnValue<kaguya::LuaTable>(lua_State*, int, int, kaguya::types::typetag<kaguya::LuaTable>) in osm_lua_processing.o
      kaguya::FunctionResults::FunctionResults(lua_State*, int, int) in osm_lua_processing.o
      ...
  "_lua_callk", referenced from:
      kaguya::detail::property_index_function(lua_State*) in osm_lua_processing.o
      kaguya::detail::property_newindex_function(lua_State*) in osm_lua_processing.o
      kaguya::class_userdata::detail::call_constructor_metamethod(lua_State*) in osm_lua_processing.o
  "_lua_copy", referenced from:
      kaguya::class_userdata::detail::call_constructor_metamethod(lua_State*) in osm_lua_processing.o
  "_lua_isinteger", referenced from:
      bool kaguya::detail::LuaVariantImpl<kaguya::LuaRef>::isType<int>() const in read_shp.o
  "_lua_pcallk", referenced from:
      kaguya::State::dofile(char const*, kaguya::LuaTable const&) in osm_lua_processing.o
      kaguya::LuaTable kaguya::detail::LuaFunctionImpl<kaguya::TableKeyReferenceProxy<char const*> >::call<kaguya::LuaTable, kaguya::LuaTable&>(kaguya::LuaTable&&&) in osm_lua_processing.o
      kaguya::State::dostring(char const*, kaguya::LuaTable const&) in osm_lua_processing.o
      kaguya::FunctionResults kaguya::detail::LuaFunctionImpl<kaguya::TableKeyReferenceProxy<char const*> >::call<kaguya::FunctionResults, OsmLuaProcessing*>(OsmLuaProcessing*&&) in osm_lua_processing.o
      kaguya::FunctionResults kaguya::detail::LuaFunctionImpl<kaguya::LuaFunction>::call<kaguya::FunctionResults, OsmLuaProcessing*>(OsmLuaProcessing*&&) in osm_lua_processing.o
  "_lua_rawgetp", referenced from:
      kaguya::lua_type_traits<kaguya::FunctionInvokerType<std::__1::tuple<void (*)(kaguya::ObjectWrapperBase*)> >, void>::invoke(lua_State*) in osm_lua_processing.o
      OsmLuaProcessing const* kaguya::get_const_pointer<OsmLuaProcessing>(lua_State*, int, kaguya::types::typetag<OsmLuaProcessing>) in osm_lua_processing.o
      OsmLuaProcessing* kaguya::get_pointer<OsmLuaProcessing>(lua_State*, int, kaguya::types::typetag<OsmLuaProcessing>) in osm_lua_processing.o
      kaguya::UserdataMetatable<OsmLuaProcessing, void>::createMatatable(lua_State*) const in osm_lua_processing.o
      kaguya::class_userdata::newmetatable(lua_State*, std::type_info const&, char const*) in osm_lua_processing.o
      bool kaguya::class_userdata::available_metatable<OsmLuaProcessing>(lua_State*) in osm_lua_processing.o
      kaguya::class_userdata::setmetatable(lua_State*, std::type_info const&) in osm_lua_processing.o
      ...
  "_lua_rawlen", referenced from:
      std::__1::vector<kaguya::LuaRef, std::__1::allocator<kaguya::LuaRef> > kaguya::detail::LuaTableOrUserDataImpl<kaguya::LuaTable>::keys<kaguya::LuaRef, std::__1::allocator<kaguya::LuaRef> >() const in read_shp.o
      std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > kaguya::detail::LuaTableOrUserDataImpl<kaguya::LuaStackRef>::values<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >() const in osm_lua_processing.o
  "_lua_rawsetp", referenced from:
      kaguya::UserdataMetatable<OsmLuaProcessing, void>::createMatatable(lua_State*) const in osm_lua_processing.o
      kaguya::class_userdata::newmetatable(lua_State*, std::type_info const&, char const*) in osm_lua_processing.o
      bool kaguya::class_userdata::available_metatable<OsmLuaProcessing>(lua_State*) in osm_lua_processing.o
      kaguya::class_userdata::setmetatable(lua_State*, std::type_info const&) in osm_lua_processing.o
  "_lua_rotate", referenced from:
      kaguya::LuaStackRef kaguya::detail::LuaTableOrUserDataImpl<kaguya::LuaStackRef>::getRawField<char [7]>(char const (&) [7]) const in osm_lua_processing.o
      kaguya::class_userdata::newmetatable(lua_State*, std::type_info const&, char const*) in osm_lua_processing.o
      bool kaguya::class_userdata::available_metatable<OsmLuaProcessing>(lua_State*) in osm_lua_processing.o
      kaguya::class_userdata::setmetatable(lua_State*, std::type_info const&) in osm_lua_processing.o
  "_lua_tointegerx", referenced from:
      kaguya::lua_type_traits<long long, void>::get_type kaguya::detail::LuaVariantImpl<kaguya::LuaRef>::get<long long>() const in read_shp.o
  "_lua_tonumberx", referenced from:
      kaguya::lua_type_traits<double, void>::get_type kaguya::detail::LuaVariantImpl<kaguya::LuaRef>::get<double>() const in read_shp.o
      kaguya::util::stackValueDump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, lua_State*, int, int) in read_shp.o
      int kaguya::nativefunction::_call_apply<void (OsmLuaProcessing::*&)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, float), OsmLuaProcessing&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, float, 1ul, 2ul, 3ul>(lua_State*, void (OsmLuaProcessing::*&&&)(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, float), kaguya::nativefunction::index_tuple<1ul, 2ul, 3ul>, kaguya::util::FunctionSignatureType<void, OsmLuaProcessing&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, float>) in osm_lua_processing.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [tilemaker] Error 1
rm include/vector_tile.pb.cc include/osmformat.pb.cc
__________________________________________________________
I guess it's some link error. I tried to mkdir build, and use CMakeList to generate Makefile, but failed

**Comment by Diniden** [comment]
Reading up on your issue and looking at the output you can troubleshoot by making sure the

`-l<path to library>` options are pointing to valid sources. Your errors are involving the dependencies to be installed so there is a good chance the system you used to install said dependencies did not place them where the library link options expected them to be.

That is at least a very good start point for this issue.

**Comment by tomquas** [comment]
fwiw, i had to modify the build to use c++14 instead of c++11 to compile against the latest lib versions - boost 1.75.0 et al on macos big sur (11.2)

**Comment by mvexel** [comment]
Different error messages, same solution. MacOS Big Sur, installed dependencies as per instructions.

```
> c++ -v
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
> make
```

gives as first error

```
/usr/local/include/boost/geometry/util/type_traits_std.hpp:54:12: error: no template named 'conditional_t' in namespace 'std'; did you mean 'conditional'?
    : std::conditional_t<Trait::value, conjunction<Traits...>, Trait>
      ~~~~~^
```

@tomquas 's solution to use `c++14` instead of `c++11` makes everything compile fine.

**Comment by systemed** [comment]
Boost::Geometry [requires C++14 as of 1.75](https://www.boost.org/users/history/version_1_75_0.html) so yes, we probably need to fall into line too.

**Comment by systemed** [comment]
Fixed in #229 - feel free to reopen if it reoccurs.
